x/evm/statedb/journal.go

```diff
@@ -335,3 +318,46 @@ func (ch accessListAddSlotChange) Revert(s *StateDB) {
 func (ch accessListAddSlotChange) Dirtied() *common.Address {
 	return nil
 }
+
+// ------------------------------------------------------
+// PrecompileSnapshotBeforeRun
+
+// PrecompileCalled: Precompiles can alter persistent storage of other
+// modules. These changes to persistent storage are not reverted by a `Revert` of
+// [JournalChange] by default, as it generally manages only changes to accounts
+// and Bank balances for ether (NIBI).
+//
+// As a workaround to make state changes from precompiles reversible, we store
+// [PrecompileCalled] snapshots that sync and record the prior state
+// of the other modules, allowing precompile calls to truly be reverted.
+//
+// As a simple example, suppose that a transaction calls a precompile.
+//  1. If the precompile changes the state in the Bank Module or Wasm module
+//  2. The call gets reverted (`revert()` in Solidity), which shoud restore the
+//     state to a in-memory snapshot recorded on the StateDB journal.
+//  3. This could cause a problem where changes to the rest of the blockchain state
+//     are still in effect following the reversion in the EVM state DB.
+type PrecompileCalled struct {
+	MultiStore store.CacheMultiStore
+	Events     sdk.Events
+}
+
+var _ JournalChange = PrecompileCalled{}
+
+// Revert rolls back the [StateDB] cache context to the state it was in prior to
+// the precompile call. Modifications to this cache context are pushed to the
+// commit context (s.evmTxCtx) when [StateDB.Commit] is executed.
+func (ch PrecompileCalled) Revert(s *StateDB) {
+	s.cacheCtx = s.cacheCtx.WithMultiStore(ch.MultiStore)
```

Comment:
Note to future self: since we only create a `cacheCtx` whenever a precompile is called, then reverting a precompile call can just get rid of the `cacheCtx` and reset the `writeToCommitCtxFromCacheCtx` function to `nil`. That would be a cleaner way to revert a `PrecompileCalled` journal entry rather than deep copy the multistore and save it in a snapshot, but I'd have to test that implementation out.

Reply:
Opening as an improvement ticket